src/cluster.c

```diff
@@ -662,7 +662,7 @@ void ShardingInfoRDBLoad(RedisModuleIO *rdb)
         RedisModule_Assert(sg != NULL);
 
         /* set local flag */
-        if (!strncmp(sg->id, rr->log->dbid, RAFT_DBID_LEN)) {
+        if (!strncmp(sg->id, rr->meta.dbid, RAFT_DBID_LEN)) {
```

Comment:
I'm confused by adding node/dbid and the like into Metadata when it still needs to be in log for log writing (see writeHeader())

Reply:
ok, I see, the point is, in the future we might have multiple log files we are juggling, so this makes a single place to find it.  With that said, why would it matter which log object we get it from?  should juggling really matter?

Reply:
> I'm confused by adding node/dbid and the like into Metadata when it still needs to be in log for log writing

Added `dbid` and `nodeid` to metadata to verify files (metadata, log and snapshot) match on restart.

Btw, normally, `dbid` and `nodeid` don't have to be in the log header. Previously, metadata was part of the log header. When I moved metadata out of the log file, I didn't delete `dbid` and `nodeid` from the log file. Just keeping `dbid` and `nodeid` in both files to verify we don't mix files of another node.

> why would it matter which log object we get it from?

It doesn't matter which log object we get it. If we leave this unchanged, it still works. 

When we restart, as the first thing, we read metadata file. From that point, we have `dbid` there. So, I'm just trying to centralize `dbid` access, move it to metadata object. If we do that, we don't need to open log file just after we read metadata file: https://github.com/RedisLabs/redisraft/blob/a7dca73ce58997a1176f1944aac31cff3b2a65c3/src/redisraft.c#L1980

Currently, we are opening log file(without reading entries) just to initialize `rr->log->dbid`.

Reply:
I'm trying to understand the motivation for moving vote and term to this file.

lets forget vote for a moment. what happens if we don't record term.

we are in term X, which should imply we have a log record that says we are in term X, so by replaying log, we will know that.

so I'm thinking it has to do with voter, and not simply what term we are in.  Is it what term we last voted in so that if we restart, we don't vote for a different node in same term?

Reply:
> we are in term X, which should imply we have a log record that says we are in term X, so by replaying log, we will know that.

If you are a follower, you may not have an entry with term X. e.g. you restart before receiving the noop entry. Now, this follower will accept RPCs with lower term. Not sure if Raft handles this well on corner cases. 

Just thinking loud, on a split brain, that follower can receive a lower term entry from leader A and add it to its log. Leader A can commit the entry. The leader with the new term, leader B, can override the entry later as it is trying to send noop entry with a higher term for the same index? 

> so I'm thinking it has to do with voter, and not simply what term we are in. Is it what term we last voted in so that if we restart, we don't vote for a different node in same term?

Yes, `vote` is here to prevent double vote on restart.

Reply:
if we are a follower, didn't we by definition get a noop for term X?  isn't that what makes us a follower?  how would we become a follower without it?  voting makes us a follower?  or does voting make us leaderless on new term (not really a follower, right?) until we get the noop?  so we would reject incoming messages from old leader, as old term?

Reply:
When you vote for a node, as you need to increment your term, you are not a follower of the previous leader anymore. 
Now, until you receive NOOP, say whatever you like :) You can say the node is a follower or leaderless. It doesn't matter for Raft as long as you reject older term RPCs. (although, paper indicates that a node becomes a follower when it receives an RPC message with a higher term, requestvote is an RPC).